Memtable flush: wait for sstable count reduction if needed

bhalevy opened the pull request:
Called from try_flush_memtable_to_sstable,
maybe_wait_for_sstable_count_reduction will wait for
compaction to catch up with memtable flush if there
the bucket to compact is inflated, having too many
sstables.  In that case we don't want to add fuel
to the fire by creating yet another sstable.

Fixes #4116

avikivity commented on replica/table.cc:
```diff
@@ -741,29 +741,34 @@ future<> table::maybe_wait_for_sstable_count_reduction() {
+    auto num_runs_with_memtable_origin = [this] {
+        auto& cm = _compaction_manager;
+        auto& cs = get_compaction_strategy();
+        auto desc = cs.get_sstables_for_compaction(as_table_state(), cm.get_strategy_control(), cm.get_candidates(*this));
```
What if this chooses sstables that aren't memtable-origin for the "best" compaction? Then we get a false negative.

Probably it's not terrible. If the best compaction happens not to include memtable-origin sstables (or not all of them) then the memtable-origin sstables probably aren't a problem.

But we may still have a problem with filtering only on memtable-origin sstables: the first compaction converts memtable-origin sstables to non-memtable-origin sstables, and if that compaction is fast enough but the following compactions aren't, we may still be creating sstables faster than we're compacting them. It's probably unlikely to happen though.

We may want to drop the whole memtable-origin stuff and instead just ask the compaction strategy if it's overloaded and if so stop flushing new memtables. Exactly what "overloaded" means will be left to the compaction strategy to decide.

We could do this later though.

bhalevy replied:
> What if this chooses sstables that aren't memtable-origin for the "best" compaction? Then we get a false negative.
> 
> Probably it's not terrible. If the best compaction happens not to include memtable-origin sstables (or not all of them) then the memtable-origin sstables probably aren't a problem.

Agreed.

> But we may still have a problem with filtering only on memtable-origin sstables: the first compaction converts memtable-origin sstables to non-memtable-origin sstables, and if that compaction is fast enough but the following compactions aren't, we may still be creating sstables faster than we're compacting them. It's probably unlikely to happen though.
> 
> We may want to drop the whole memtable-origin stuff and instead just ask the compaction strategy if it's overloaded and if so stop flushing new memtables. Exactly what "overloaded" means will be left to the compaction strategy to decide.
> 
> We could do this later though.

The "compaction overloaded" metric is similar to the compaction backlog.
The difference is that the backlog still doesn't take into account the number of sstables, only their size (don't be mislead by #4076 being closed), so we don't factor in potential read-amplification.

avikivity replied:
It doesn't have to be similar to compaction backlog. It can be a pure sstable count thing.

avikivity commented on replica/table.cc:
```diff
@@ -741,29 +741,34 @@ future<> table::maybe_wait_for_sstable_count_reduction() {
     using namespace std::chrono_literals;
     auto start = db_clock::now();
-    co_await _sstables_changed.wait(sstable_count_below_threshold);
+    co_await _sstables_changed.wait([&num_runs_with_memtable_origin, threshold] { return num_runs_with_memtable_origin() <= threshold; });
```
We're mixing three different things here: memtable-origin filtering, the compaction strategy's choice of the best compaction, and the threshold. So it's possible we're just moving the edge cases around,

bhalevy replied:
From the test passing vs. hanging before there is a bit of evidence it's doing a little more than merely moving the edge cases around :)

The gist of the change is not counting the overall number of memtable-origin sstables as they might be scattered across buckets (as was in the TWCS test where each flush generated a single sstable in each time window), and then waiting for their compaction was futile because they didn't trigger compaction.

Focusing on the "best" bucket + plus having the threshold derived from `max_compaction_threshold` ensures compaction is needed for the bucket.

Filtering by memtable-origin may indeed be dropped here so that the logic would detect buckets that are inflated for any reason, not only due to memtable flush.

bhalevy replied:
Also, focusing on the "best" bucket ensures that compaction for it will be imminent.
If other buckets are inflated too (like in the TWCS test example, where all windows are filled at the same rate and therefore will cross the threshold at the same time), we should wait for them one at a time at the loop below.

avikivity replied:
> From the test passing vs. hanging before there is a bit of evidence it's doing a little more than merely moving the edge cases around :)

It's moving the edge cases from excluding that test to including that test. But it's hard to predict what cases are excluded now because too many things are in motion here.

> 
> The gist of the change is not counting the overall number of memtable-origin sstables as they might be scattered across buckets (as was in the TWCS test where each flush generated a single sstable in each time window), and then waiting for their compaction was futile because they didn't trigger compaction.

What if some compaction strategy selects candidates that aren't memtable-origin as the best set for some reason, but there's still lots of them?

Maybe it can be argued that the compaction strategy is wrong. But it's still an indirect way.

Maybe memtable-origin are sorted into 10 different buckets, so now the total number of memtable-origin sstables is 10*threshold? Still not a disaster.

The problem is that it's all indirectly tied together. From what we know of specific compaction strategies it should work, but the APIs themselves don't guarantee it.

> 
> Focusing on the "best" bucket + plus having the threshold derived from `max_compaction_threshold` ensures compaction is needed for the bucket.
> 
> Filtering by memtable-origin may indeed be dropped here so that the logic would detect buckets that are inflated for any reason, not only due to memtable flush.

Maybe that's sensible. If compaction is not keeping up (maybe the controller is wrong, or someone set static shares, or someone set fixed bandwidth) then we don't make the problem worse by not creating more sstables. It's drastic (service stops) but just ignoring the problem is bad too.

avikivity commented on replica/table.cc:
```diff
@@ -764,7 +764,14 @@ future<> table::maybe_wait_for_sstable_count_reduction() {
+    try {
+        auto wait_duration = (1000ms * count) / threshold;
+        co_await _sstables_changed.wait(wait_duration, [&num_runs_with_memtable_origin, threshold] { return num_runs_with_memtable_origin() <= threshold; });
+    } catch (const condition_variable_timed_out&) {
```
Doesn't this break the condition? If we're slow, we can still end up with an unbootable node now.

bhalevy replied:
Never mind, we can just rely on the bucket-based condition and checking that regular compaction is possible and wait until the condition is satisfied with no timeout.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1149/

bhalevy commented:
In v2 (bf449229de)
- "table: stop: break _sstables_changed condition variable"
  - swallow `broken_condition_variable` exception in `maybe_wait_for_sstable_count_reduction`
  - and go on with flushing
- dropped "table: maybe_wait_for_sstable_count_reduction: limit wait timeout"

bhalevy commented on replica/table.cc:
```diff
@@ -739,17 +739,22 @@ future<> table::maybe_wait_for_sstable_count_reduction() {
-            | boost::adaptors::transformed(std::mem_fn(&sstables::sstable::run_identifier)));
-        const auto threshold = std::max(schema()->max_compaction_threshold(), 32);
-        return sstable_runs_with_memtable_origin.size() <= threshold;
+            | boost::adaptors::transformed([] (const sstables::shared_sstable& sst) {
```
The member function is for class sstable.
I'm not even sure how it ran against a shared_sstable (which is lw_shared_ptr<sstable>).
Maybe the compiler created the shim between the too.
Anyhow, I think the lambda function is clearer in this case.

mikolajsieluzycki replied:
From https://en.cppreference.com/w/cpp/utility/functional/mem_fn  `Both references and pointers (including smart pointers) to an object can be used. ` It's purely stylistic and I have no strong opinion here.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -925,6 +925,19 @@ void compaction_manager::submit(replica::table* t) {
+    if (!could_lock) {
+        return false;
+    }
+    lk.read_unlock();
```
1. this is all synchronous code so there's no preemption between the try_read_lock call to read_unlock.
2. and even if we would yield, read_unlock just returns the single unit that read_lock acquired back to the semaphore, so if another fiber was able to acquire a read_lock in between, read_unlock wouldn't unlock the lock for writers yet.

bhalevy commented:
In v3 (14933acb5), added following patches:
- 1f16a4b59 table: maybe_wait_for_sstable_count_reduction: count all runs for compaction
- 01f05266a compaction_manager: move maybe_wait_for_sstable_count_reduction logic from table
- 14933acb5 table: get rid of _sstables_changed

raphaelsc commented on replica/table.cc:
```diff
@@ -779,6 +781,7 @@ table::stop() {
     }
     return _async_gate.close().then([this] {
         return await_pending_ops().finally([this] {
+            _sstables_changed.broken();
```
conditiona_variable::broken() doc:
    /// Signal to waiters that an error occurred.  \ref wait() will see
    /// an exceptional future<> containing the provided exception parameter.
    /// The future is made available immediately.

isn't it better to just call signal()?

bhalevy replied:
No, because the condition will not evaluate to true.
We'll need to check for the stop condition in the functor.
Breaking the condition variable works like semaphore::broken or gate::close in this sense

raphaelsc replied:
Thanks for the explanation.

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -486,6 +486,7 @@ public:
     friend class compacting_sstable_registration;
     friend class compaction_weight_registration;
     friend class compaction_manager_test;
+    friend class replica::table;
```
Yes.  I can rearrange the seies to get rid of 5he inrermediate patches once we reach agreement on the end result

bhalevy commented on replica/table.cc:
```diff
@@ -739,17 +739,22 @@ future<> table::maybe_wait_for_sstable_count_reduction() {
     if (_async_gate.is_closed() || is_auto_compaction_disabled_by_user()) {
```
I don't think they share their run id.
I saw ~60 sstables with unique run id created every flush.

bhalevy replied:
How can they share a run id if they are not disjoint?
I.e. the same partition can be present in different time windows after segregation.

raphaelsc replied:
That was from the top of my head. For flush we're indeed not reusing the run id, but I think that's only affecting segregation on compaction. I will fix that. It shouldn't be causing any issues today, but it's a bug and must be fixed.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -925,6 +925,19 @@ void compaction_manager::submit(replica::table* t) {
+bool compaction_manager::can_perform_regular_compaction(replica::table* t) {
+    if (!can_proceed(t) || t->is_auto_compaction_disabled_by_user()) {
+        return false;
+    }
```
@bhalevy could you please explain a bit more on the motivation for this patch? do we need to mimick the locking for submitting a regular compaction job? I think that will be hard to keep in sync over time. for example, we also have the weight serialization which can prevent a job from kicking in. isn't it enough to have the timer if we decided that we should block flushes? if L0 sstables is huge, then it's not a bad idea to wait for time T before proceeding with flush, that can give compaction some time to keep up.

bhalevy replied:
My main worry is what happens during major compaction.
It might me tied up for a long time compacting the high tiers.

I think that on the long term we may want to allow regular compaction of new sstable during major compaction of the existing sstable set.  Plus, start a targeted compaction job for the inflated buckets maybe even with higher shares than refular compaction to reduce the number of sstables, ir alternatively make sure regular compaction picks those buckets next time it considers sstables for compaction (and wait for it)

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -925,6 +925,19 @@ void compaction_manager::submit(replica::table* t) {
 }
 
+bool compaction_manager::can_perform_regular_compaction(replica::table* t) {
+    if (!can_proceed(t) || t->is_auto_compaction_disabled_by_user()) {
```
I think it'd be better moved to the compaction manager, kept in the table state.

raphaelsc commented on replica/table.cc:
```diff
@@ -739,21 +739,18 @@ future<> table::maybe_wait_for_sstable_count_reduction() {
             desc.sstables
-            | boost::adaptors::filtered([] (const sstables::shared_sstable& sst) {
-                return sst->get_origin() == "memtable";
-            })
```
That's a big change :-) But I think it's the right thing to do. If any bucket has severely fallen behind, then it means read amplification is terrible and LSM tree needs to stop ingesting more data, so compaction can keep up and maintain read amplification within bounds. Compaction strategy should strive to return the bucket which is causing most of the backlog, if any strategy isn't doing that, then it is its fault. 

LCS returns L0 if it's falling behind
STCS and ICS returns any bucket with most files
So does TWCS, which is essentially a set of LSM trees.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -938,6 +940,43 @@ bool compaction_manager::can_perform_regular_compaction(replica::table* t) {
 
+future<> compaction_manager::maybe_wait_for_sstable_count_reduction(replica::table *t) {
+    if (!can_perform_regular_compaction(t)) {
+        co_return;
```
I am thinking that major perhaps shouldn't completely block regular compactions. it should take write lock, to wait for ongoing regular, but as soon as it takes write lock, then perhaps should switch to read lock, to allow compaction on newly created files, to not affect read amplificaiton on L0.

raphaelsc replied:
It's terrible really that major blocks compaction on newly created files. I think that's a very important thing to fix. Will open an issue.

raphaelsc replied:
issue: https://github.com/scylladb/scylla/issues/10961

raphaelsc commented on compaction/compaction_manager.hh:
```diff
@@ -292,6 +292,7 @@ private:
 
     class strategy_control;
     std::unique_ptr<strategy_control> _strategy_control;
+    condition_variable _compaction_done;
```
can two tables safely share the condition_variable? isn't it better to make it per table in compaciton_manager::compaction_state ?

bhalevy replied:
good catch.  They can, but t doesn't help the waiter if it's being signaled when compaction is done on another table.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1160/

bhalevy commented:
In v4 (de6953dad1)
- Moved logic to compaction manger earlier in the series
- made compaction_done condition variable a member of compaction_state
  - signaled on `finish_compaction`
  - broken when state is destroyed (in case we missed finish compaction before destroying the state)
- mentioned #10961

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1169/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1169/

Failed on seemingly unrelated issue.  Rebuilding...

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1176/

raphaelsc commented on compaction/compaction_manager.hh:
```diff
@@ -67,6 +67,13 @@ private:
         // Raised by any function running under run_with_compaction_disabled();
         long compaction_disabled_counter = 0;
 
+        // Signaled whenever a compaction task compeletes.
```
typo: compeletes -> completes

raphaelsc reviewed (approved):
very good. only had a comment on typo, but can be fixed later.

bhalevy commented:
In v5 (781b2004bd)
- Fixed typo in comment
- Fixed inadvertent update of the seastar and tools/python3 submodules

bhalevy commented:
I'll rebase on top of https://github.com/scylladb/scylla/commit/419fe652598f01fd0cfc19d642313bbb5380f630

bhalevy commented:
In v6 (2aeef1f1f6e737f1a817ea4c376d31da5db4a77b):
- Rebased on top of 419fe652598f01fd0cfc19d642313bbb5380f630

bhalevy commented:
Please do not merge.
I need to fix an issue found by the `test_compaction_removes_ttld_data_by_time_windows`
that's caused by calling `get_sstables_for_compaction` from the flush path, since it's not fully idempotent.
It assumes the descriptor it returns is always going to be compacted which is not the case here and next time it is called it fails to find fully-expired sstables it found in the previous call since `_last_expired_check` was already updated.

bhalevy commented:
Related to the above: https://github.com/scylladb/scylla/issues/10989

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1221/

bhalevy commented:
Depends on https://github.com/scylladb/scylla/pull/10990

raphaelsc commented:
> I think it makes the compaction strategy implementation a bit more complicated, but not terribly so.

@bhalevy  TWCS internally has a notion of "now " as the timestamp of newest window. I am thinking if we could play with it to give a sense of idempotency to the expired check.

Something like:

```
diff --git a/compaction/time_window_compaction_strategy.cc b/compaction/time_window_compaction_strategy.cc
index 268834256..95f4ba0c2 100644
--- a/compaction/time_window_compaction_strategy.cc
+++ b/compaction/time_window_compaction_strategy.cc
@@ -224,10 +224,13 @@ time_window_compaction_strategy::get_sstables_for_compaction(table_state& table_
     // Find fully expired SSTables. Those will be included no matter what.
     std::unordered_set<shared_sstable> expired;
 
-    if (db_clock::now() - _last_expired_check > _options.expired_sstable_check_frequency) {
+    timestamp_type now = api::timestamp_clock::now().time_since_epoch().count();
+    auto expired_check_frequency = std::chrono::duration_cast<api::timestamp_clock::duration>(_options.expired_sstable_check_frequency).count();
+
+    if (now - _last_expired_check > expired_check_frequency) {
         clogger.debug("TWCS expired check sufficiently far in the past, checking for fully expired SSTables");
         expired = table_s.fully_expired_sstables(candidates, compaction_time);
-        _last_expired_check = db_clock::now();
+        _last_expired_check = _highest_window_seen;
     } else {
         clogger.debug("TWCS skipping check for fully expired SSTables");
     }

```

raphaelsc commented:
> @raphaelsc let's consider that as a follow up. It seems a bit risky to me. I'll push a simple version that just doesn't update `_last_expired_check` in dry_run mode.

fair enough.

what about

```
auto cs = make_compaction_strategy(_schema->compaction_strategy(), _schema->compaction_strategy_options());
cs->get_sstables_for_compaction(...)
```

?

so this gets the dry run behavior without the additional param.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -863,7 +863,7 @@ class compaction_manager::regular_compaction_task : public compaction_manager::t
             }
-            if (!_cm.can_register_compaction(&t, weight, descriptor.fan_in())) {
+            if (!descriptor.has_only_fully_expired && !_cm.can_register_compaction(&t, weight, descriptor.fan_in())) {
                 cmlog.debug("Refused compaction job ({} sstable(s)) of weight {} for {}.{}, postponing it...",
```
Actually the patch below is better:

```
@@ -167,6 +167,11 @@ bool compaction_manager::can_register_compaction(replica::table* t, int weight,
     // TODO: Maybe allow only *smaller* compactions to start? That can be done
     // by returning true only if weight is not in the set and is lower than any
     // entry in the set.
+
+    // Weightless compaction doesn't have to be serialized, and won't dillute overall efficiency.
+    if (!weight) {
+        return true;
+    }
```

Fixes the problem you're observing.

bhalevy replied:
calculate_weight simply returns 0 for `has_only_fully_expired`, but `can_register_compaction` doesn't special-case weight 0, and it serializes it like any other weight.
Note that `max_ongoing_compaction_test` relies on that and it explicitly checked that these compactions are serialized.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -164,6 +164,10 @@ bool compaction_manager::can_register_compaction(replica::table* t, int weight,
     if (!t->get_compaction_strategy().parallel_compaction() && has_table_ongoing_compaction(t)) {
```
Hi @bhalevy. perhaps we can remove this commit from this branch and leave it for https://github.com/scylladb/scylla/pull/10990 which is dedicated to fixing the problem described. what's your plan?

bhalevy replied:
It is part of https://github.com/bhalevy/scylla/commits/always-register-fully-expired-sstables-for-compaction
I made this PR a draft until the dependencies are resolved.
Currently it has only the last 2 unique commits.

raphaelsc replied:
thanks for the explanation.

bhalevy commented:
Note: let's run this PR through SCT to make sure it doesn't cause regressions.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -956,6 +961,64 @@ void compaction_manager::submit(replica::table* t) {
+    auto could_lock = lk.try_read_lock();
+    if (!could_lock) {
+        return false;
+    }
```
now that regular can proceed in parallel to major, perhaps we can remove this code which mimicks the serialization logic?

bhalevy replied:
agreed

bhalevy commented:
SCT run: https://jenkins.scylladb.com/view/master/job/scylla-master/view/daily-master-longevity-tests/job/longevity/job/longevity-10gb-3h-gce-test/212/

avikivity commented:
> In v7 ([2f2dcf7](https://github.com/scylladb/scylla/commit/2f2dcf7773bc0b5b0675477c6e954bcd06b3a9a7))
> 
> * rebased on top of [6332816](https://github.com/scylladb/scylla/commit/6332816ccf2ce637a65a20bfa810bf81441c28f4)
> * compaction_manager: add maybe_wait_for_sstable_count_reduction
>   
>   * make a new compaction_strategy object so that any side effects of get_sstables_for_compaction will not be applied on the table's comapction_strategy instance


That seems very drastic. Isn't it expensive? Just to check if we can create a new sstable?

bhalevy commented:
> That seems very drastic. Isn't it expensive? Just to check if we can create a new sstable?

It shouldn't be too expensive.
But I have an idea how to make twcs::get_sstables_for_compaction idempotent inexpensively.
I think it would be better to do it this way.

raphaelsc commented:
> > In v7 ([2f2dcf7](https://github.com/scylladb/scylla/commit/2f2dcf7773bc0b5b0675477c6e954bcd06b3a9a7))
> > 
> > * rebased on top of [6332816](https://github.com/scylladb/scylla/commit/6332816ccf2ce637a65a20bfa810bf81441c28f4)
> > * compaction_manager: add maybe_wait_for_sstable_count_reduction
> >   
> >   * make a new compaction_strategy object so that any side effects of get_sstables_for_compaction will not be applied on the table's comapction_strategy instance
> 
> That seems very drastic. Isn't it expensive? Just to check if we can create a new sstable?

It's allocation for compaction_strategy object + some basic option parsing. Given that it happens on flush interval only, is it a problem?. The alternative solution is adding a dry run param, so to indicate no internal state should be updated, but that may introduce maintenance complexity.

bhalevy commented:
> The alternative solution is adding a dry run param, so to indicate no internal state should be updated, but that may introduce maintenance complexity.

Not only, please look at my patch in the next version.
I think it's pretty straight forward to make it idempotent without a dry_run parameter.

raphaelsc commented on compaction/time_window_compaction_strategy.cc:
```diff
@@ -221,22 +221,24 @@ time_window_compaction_strategy::get_sstables_for_compaction(table_state& table_
+        if (!expired.empty()) {
+            clogger.debug("[{}] Going to compact {} expired sstables", fmt::ptr(this), expired.size());
+            return compaction_descriptor(has_only_fully_expired::yes, std::vector<shared_sstable>(expired.begin(), expired.end()), service::get_local_compaction_priority());
+        }
```
Good. It's an improvement even without the wait behavior introduced in the series.

raphaelsc reviewed (approved):
Looks good to me

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1359/

denesb commented on compaction/time_window_compaction_strategy.cc:
```diff
@@ -225,19 +225,20 @@ time_window_compaction_strategy::get_sstables_for_compaction(table_state& table_
 
     if (db_clock::now() - _last_expired_check > _options.expired_sstable_check_frequency) {
-        clogger.debug("TWCS expired check sufficiently far in the past, checking for fully expired SSTables");
+        clogger.debug("[{}] TWCS expired check sufficiently far in the past, checking for fully expired SSTables", fmt::ptr(this));
```
How is logging `this` help? I thought there is just one of these per table.

bhalevy replied:
IF we always use that one by reference.
The debugging code helps if the compaction strategy object is copied and the side effects in this function happen on a temporary copy rather than on the authoritative instance.

denesb commented on compaction/time_window_compaction_strategy.cc:
```diff
@@ -221,22 +221,24 @@ time_window_compaction_strategy::get_sstables_for_compaction(table_state& table_
+
+        // Find fully expired SSTables. Those will be included no matter what.
+        std::unordered_set<shared_sstable> expired;
         expired = table_s.fully_expired_sstables(candidates, compaction_time);
```
Nit: might as well change to `auto expired = ...`

bhalevy replied:
This is done in a following patch.
I implemented it this way in this patch just to minimize the diff and make the logic changes clearer.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -495,6 +495,10 @@ inline compaction_controller make_compaction_controller(compaction_manager::sche
     return compaction_controller(csg, 250ms, std::move(fn));
```
s/infalted/inflated/

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -956,6 +961,50 @@ void compaction_manager::submit(replica::table* t) {
+        auto desc = cs.get_sstables_for_compaction(t->as_table_state(), get_strategy_control(), get_candidates(*t));
+        return boost::copy_range<std::unordered_set<utils::UUID>>(
+            desc.sstables
+            | boost::adaptors::transformed(std::mem_fn(&sstables::sstable::run_identifier))).size();
```
This is a weird way to count something. There got to be some `std::acumulate` or similar that is a better fit.

denesb replied:
I missed the the purpose of the set, indeed it doesn't seem like accumulate() would work here.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -956,6 +961,50 @@ void compaction_manager::submit(replica::table* t) {
+    submit(t);
+    using namespace std::chrono_literals;
+    auto start = db_clock::now();
+    auto& cstate = get_compaction_state(t);
```
Maybe `submit(t)` should also return a reference to the compaction state, for convenience.

bhalevy replied:
This is going to change soon with @raphaelsc's on-going work so submit will be called with the table_state rather than with the table* so I'd rather not do this change here.

denesb replied:
ok

denesb commented on replica/table.cc:
```diff
@@ -688,6 +688,9 @@ table::try_flush_memtable_to_sstable(lw_shared_ptr<memtable> old, sstable_write_
                 _memtables->erase(old);
                 co_return stop_iteration::yes;
             }
+            if (!_async_gate.is_closed() || is_auto_compaction_disabled_by_user()) {
```
Shouldn't this be `!is_auto_compaction_disabled_by_user()`?

bhalevy replied:
almost... :)
It should '&& !is_auto_compaction_disabled_by_user' (or !(_async_gate.is_closed() || is_auto_compaction_disabled_by_user()) which was the original intention)

good catch

bhalevy replied:
I'll just delete this term, since it's checked again anyhow by `maybe_wait_for_sstable_count_reduction`

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -353,7 +353,7 @@ class compaction_manager::major_compaction_task : public compaction_manager::tas
         // candidates are sstables that aren't being operated on by other compaction types.
         // those are eligible for major compaction.
         auto* t = _compacting_table;
-        sstables::compaction_strategy cs = t->get_compaction_strategy();
```
I don't understand what is the connection between idempotency and being able to return something by reference.

bhalevy replied:
`table::get_compaction_strategy` returns a reference to its compaction_strategy instance, but it was copied here - causing side effects to be done on the copy rather than on the original.

Looks like unintentionally, this protected against #10989, and it can be done regardless of this series, after e1e4a73793c50c9f9aa0ee0fcf0b484bcca24b2d.

Without this patch twcs get_sstables_for_compaction might be a bit less efficient as it will keep looking for fully expired sstables once `now - _last_expired_check > _options.expired_sstable_check_frequency`.

I'll file an issue about that and submit this patch in that context so we can decide whether we want to backport it there.

bhalevy commented:
In v9 (draft, 86d5a32883955595f12ff71c5660614ab7d0e14a):
- based on https://github.com/scylladb/scylla/pull/11052
- compaction_manager: add maybe_wait_for_sstable_count_reduction
  - fixed typo in changelog
  - removed `is_auto_compaction_disabled_by_user` term (as it's checked by compaction_manager)

bhalevy commented:
In v10 (145050668b):
- rebase (without #11052 that was retracted)
- dropped "compaction_manager: do not copy the table compaction_strategy object"
  - it was a red herring, will submit a followup PR to clean this compaction_strategy mess up, but it turned out to be cosmetic only.

raphaelsc reviewed (approved):
LGTM

bhalevy commented:
SCT run with latest packages:
gs://upload.scylladb.com/bhalevy/packages/pr-10954/5.1.dev-0.20220724.145050668bc0/
https://jenkins.scylladb.com/view/master/job/scylla-master/view/daily-master-longevity-tests/job/longevity/job/longevity-10gb-3h-test/529/

avikivity commented:
cosmetic: add maybe_wait_for_sstable_count_reduction doesn't give an indication that something was done (a function was added, but then what?)

raphaelsc commented:
> cosmetic: add maybe_wait_for_sstable_count_reduction doesn't give an indication that something was done (a function was added, but then what?)

change can be postponed, I think. Can we merge this?

denesb commented:
> > cosmetic: add maybe_wait_for_sstable_count_reduction doesn't give an indication that something was done (a function was added, but then what?)
> 
> change can be postponed, I think. Can we merge this?

Avi referred to the PR title, which cannot be updated in a follow-up. But the point is moot as Benny already updated it.

raphaelsc commented:
> > > cosmetic: add maybe_wait_for_sstable_count_reduction doesn't give an indication that something was done (a function was added, but then what?)
> > 
> > 
> > change can be postponed, I think. Can we merge this?
> 
> Avi referred to the PR title, which cannot be updated in a follow-up. But the point is moot as Benny already updated it.

Oh, I thought he was referring to the interface which doesn't give an indication if the waiting actually took place :-) But that's not needed as maybe_wait_for_sstable_count_reduction() dumps a warning message informing the amount of time flush had to wait.

@bhalevy perhaps we should add a metric for current # of waiters, so one can correlate the waiting of sstable reduction with any other event.

denesb commented:
Dequeued, doesn't build.

bhalevy commented:
> Dequeued, doesn't build.

I'll rebase and check that it build in the dbuild environment

bhalevy commented:
In e0c6e1ef3c: rebased on top of [#11028](https://github.com/scylladb/scylla/pull/11028)
- no functional changes (mostly passing `table_state&` rather than `table*`)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1563/
